### DISCUSSION TITLE
Spakle updates not trigger in native (rust) app

### DISCUSSION THREAD

**Discussion by hankthetank27** [hydraulic-software/conveyor#143] [Q&A] [answered]
Hi,

I've just installed and configured Conveyor to distribute an application I've been working on. While I've found it fantastic so far, I can't seem to get the Sparkle updates triggering on my MacOS build. I have tried locally, following the instructions [here](https://conveyor.hydraulic.dev/16.0/tutorial/hare/native/#create-a-sample-project), using cargo rather than cmake to build my own app instead of the example one.  When I launch the old version of the app, the sparkle update window never opens or starts. 

Unfortunately I'm also having quite a a lot of trouble getting the example app to compile so I don't have much of a point of reference. (perhaps besides the point but the example app's `make install` erroring tryin to import headers from mingw-w64. I'm not familiar with cmake, or c++ for that matter and use mingw-w64 to cross compile rust to windows. I tried to used flags/environment variables with cmake to try and use clang to compile but still getting the same errors).

I've also deployed two versions to test live which can be seen [on this repo](https://github.com/hankthetank27/edl-gen). Updates aren't triggering here either of course but maybe the configuration of this repo and the two released version can provide some more insight.

Everything works great besides the Sparkle auto updates and I'm really not sure how to debug it.

**Comment by mikehearn** [comment]
First step is to open `Console.app` and run the program. You may need to filter by process, but should see some logging statements from Conveyor or Sparkle. If they appear but update prompts don't occur, there's some config issue with Sparkle. If they don't appear at all, it means `libconveyor` isn't being loaded at all or it is being loaded but doesn't think the app is a Cocoa app. If you're writing in Rust, is it possible that the problem is the latter? The docs maybe aren't clear on this point but "native app" actually means "app that uses the native Mac APIs to at least some extent".

The way this works for native apps is that Conveyor injects a dependency into the MachO headers of the binary to force the dynamic linker to load `libconveyor` which then in a constructor function schedules some work onto the main thread's dispatch queue. This then initializes Sparkle. If your app never starts up a Cocoa event loop, Sparkle updates won't run.

**Reply by hankthetank27** [reply]
Thank you for this!

I do see a logging statement originating from `libconveyor` "Sparkle Framework is not available, no updates will occur." The app does not use Cacao or AppKit so I suppose this is the culprit.

**Reply by mikehearn** [reply]
Actually that means the Sparkle.framework itself wasn't found. Can you share a copy of the output of `conveyor make unnotarized-mac-zip` ?

**Reply by hankthetank27** [reply]
with my config set at `machines = [mac.aarch64, mac.amd64, windows.amd64]` I got an error "Task unnotarized-mac-zip is ambiguous. Try specifying e.g. -Kapp.machines=mac.amd64"

I changed the config to `machines = mac.aarch64` and got the resulting archive -
[edl-gel-1-mac-aarch64.zip](https://github.com/user-attachments/files/18100976/edl-gel-1-mac-aarch64.zip)

**Reply by mikehearn** [reply, marked as answer]
Thanks. The issue is you're missing an rpath for the Frameworks directory. You need to adjust your build system to add the right linker flags.

See here: https://conveyor.hydraulic.dev/16.0/configs/native-apps/#building-compatible-binaries

Conveyor should really detect this case and warn you, though.

**Reply by hankthetank27** [reply]
Fantastic! Seems to work now :)

For anyone else who finds themselves here the fix was adding to `build.rs`
```rust
if cfg!(target_os = "macos") {
     println!("cargo:rustc-link-arg=-Wl,-rpath,@executable_path/../Frameworks");
     println!("cargo:rustc-link-arg=-Wl,-headerpad,0xFF");
}
```
